clang/utils/TableGen/NeonEmitter.cpp

```diff
@@ -2581,6 +2581,60 @@ void NeonEmitter::runVectorTypes(raw_ostream &OS) {
   OS << "typedef double float64_t;\n";
   OS << "#endif\n\n";
 
+  OS << R"(
+typedef uint64_t fpm_t;
+
+enum __ARM_FPM_FORMAT { __ARM_FPM_E5M2, __ARM_FPM_E4M3 };
+
+enum __ARM_FPM_OVERFLOW { __ARM_FPM_INFNAN, __ARM_FPM_SATURATE };
+
+static __inline__ fpm_t __attribute__((__always_inline__, __nodebug__))
+__arm_fpm_init(void) {
+  return 0;
+}
+
+static __inline__ fpm_t __attribute__((__always_inline__, __nodebug__))
+__arm_set_fpm_src1_format(fpm_t __fpm, enum __ARM_FPM_FORMAT __format) {
+  return (__fpm & ~7ull) | (fpm_t)__format;
+}
+
+static __inline__ fpm_t __attribute__((__always_inline__, __nodebug__))
+__arm_set_fpm_src2_format(fpm_t __fpm, enum __ARM_FPM_FORMAT __format) {
+  return (__fpm & ~0x38ull) | ((fpm_t)__format << 3u);
+}
+
+static __inline__ fpm_t __attribute__((__always_inline__, __nodebug__))
+__arm_set_fpm_dst_format(fpm_t __fpm, enum __ARM_FPM_FORMAT __format) {
+  return (__fpm & ~0x1c0ull) | ((fpm_t)__format << 6u);
+}
+
+static __inline__ fpm_t __attribute__((__always_inline__, __nodebug__))
+__arm_set_fpm_overflow_mul(fpm_t __fpm, enum __ARM_FPM_OVERFLOW __behaviour) {
+  return (__fpm & ~0x4000ull) | ((fpm_t)__behaviour << 14u);
+}
+
+static __inline__ fpm_t __attribute__((__always_inline__, __nodebug__))
+__arm_set_fpm_overflow_cvt(fpm_t __fpm, enum __ARM_FPM_OVERFLOW __behaviour) {
+  return (__fpm & ~0x8000ull) | ((fpm_t)__behaviour << 15u);
+}
+
+static __inline__ fpm_t __attribute__((__always_inline__, __nodebug__))
+__arm_set_fpm_lscale(fpm_t __fpm, uint64_t __scale) {
+  return (__fpm & ~0x7f0000ull) | (__scale << 16u);
```

Comment:
Does `__scale` not need to be masked by `0x7full` to prevent a user from writing past bit 22 here?

Reply:
The `__scale` must fall within a certain range, otherwise the returns value is indeterminate.

> The valid range for the `scale` parameter is [-128, 127], inclusive. [different `scale`]
> 
> Passing an out of range argument to a helper intrinsic results in the intrinsic
> returning an indeterminate value. Passing such an indeterminate value as
> an argument to an FP8 intrinsic results in undefined behavior.

So, no, we don't **have** to mask the inputs on any of the helper intrinsics.

Now, the question is **should** we?
IMHO, we should not:
* that won't guarantee we'd obtain the value that the user actually meant
* that would hide the error of passing invalid arguments by obtaining an in-range "random" value, whereas by overwriting bits we're not supposed to, we propagate the error is give a chance for it to be detected elsewhere.

Reply:
Agreed. Having seen the description again, I don't think anyone would expect this value to be masked.

Reply:
For the `(uint32_t)__fpm`, true, we are clearing bits, which are already supposed to be zero, so in a sense we're "hiding an error", so contradicting what I just wrote, but then it's more efficient :D

Reply:
True, I'll leave it to your judgement! - Though there is also an argument for consistent behavior.